jwt/keyset.go

```diff
@@ -0,0 +1,217 @@
+package jwt
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"net/http"
+
+	"github.com/coreos/go-oidc"
+	"github.com/hashicorp/go-cleanhttp"
+	"golang.org/x/oauth2"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+// KeySet represents a set of keys that can be used to verify the signatures of JWTs.
+// A KeySet is expected to be backed by a set of local or remote keys.
+type KeySet interface {
+
+	// VerifySignature parses the given JWT, verifies its signature, and returns the claims in its payload.
+	VerifySignature(ctx context.Context, token string) (claims map[string]interface{}, err error)
+}
+
+// OIDCDiscoveryKeySet verifies JWT signatures using keys obtained by the OIDC discovery mechanism.
+type OIDCDiscoveryKeySet struct {
+	provider *oidc.Provider
+}
+
+// JSONWebKeySet verifies JWT signatures using keys obtained from a JWKS URL.
+type JSONWebKeySet struct {
+	remoteJWKS oidc.KeySet
+}
+
+// StaticKeySet verifies JWT signatures using local PEM-encoded public keys.
```

Comment:
Isn't just "public keys"?   Also, why not use the []crypto.PublicKey vs []interface{}?

Reply:
Yep, I plan to change this to just "public keys" as we discussed. I originally used PEMs because that's the way it was in the Vault plugin. I think it makes more sense to accept `[]crypto.PublicKey` directly now that this is in a library. I also like the idea of providing a helper function to parse PEMs.

Reply:
I've changed this to accept `[]crypto.PublicKey` in d944a7e. Let me know your thoughts.

Reply:
I like it.  ty!